File: C#/Structures/BST.cs
using System;

class BST{
    class Node{
        public int data;
        public Node left = null;
        public Node right = null;
        public Node(int data){
            this.data = data;
        }
    }

    Node m_root = null;
    int m_size = 0;

    public BST(){}

    
    public void insert_node(Node node){
        if(this.isEmpty()){
            this.m_root = node;
        } else{
            Node ptr = m_root;
            while (ptr != null) {
                if (node.data < ptr.data)
                    ptr = ptr.left;
                else if (node.data > ptr.data)
                    ptr = ptr.right;
                else
                    return;
            }
            if (node.data < ptr.data)
                ptr.left = node;
            if (node.data > ptr.data)
                ptr.right = node;          
        }
        ++ this.m_size;
    }
    
    public void insert_element(int data){
        Node node = new Node(data);
        this.insert_node(node);
    }
    
    public void remove_node(Node node){
        this.delete_element(node, node.data);
        -- this.m_size;
    }
    
    public void remove_element(int data){
        Node node = new Node(data);
        this.delete_element(node, data);
        -- this.m_size;
    }
    
    public void traverse_inorder(Node node){
        if(node != null){
            this.traverse_inorder(node.left);
            Console.WriteLine(node.data + " ");
            this.traverse_inorder(node.right);
        }
    }
    
    public void traverse_preorder(Node node){
        if(node != null){
            Console.WriteLine(node.data + " ");
            this.traverse_preorder(node.left);
            this.traverse_preorder(node.right);
        }
    }
    
    public void traverse_postorder(Node node){
        if(node != null){
            this.traverse_postorder(node.left);
            this.traverse_postorder(node.right);
            Console.WriteLine(node.data + " ");
        }
    }
    
    public Node root(){return this.m_root; }

    public Node find_min(Node node){
        Node ptr = node;
        while(ptr != null)
            ptr = ptr.left;
        return ptr;
    }
    
    public int size(){return this.m_size; }

    public bool is_element(int data){}

    private bool isEmpty(){ return this.m_root == null; }

    private Node delete_element(Node ptr, int data){
        if(ptr == null)
            return ptr;
        else if(data < ptr.data)
            ptr.left = this.delete_element(ptr.left, data)
        else if(data > ptr.data)
            ptr.right = this.delete_element(ptr.right, data)
        else{
            // case 1
            if(ptr.left == null && ptr.right == null)
                ptr = null
            // case 2
            else if(ptr.left == null){
                Node temp = ptr;
                ptr = ptr.right;
                temp = null;
            }
            else if(ptr.right == null){
                Node temp = ptr;
                ptr = ptr.left;
                temp = null;
            }
            // case 3
            else{
                Node temp = this.find_min(ptr.right);
                ptr.data = temp.data;
                ptr.right = delete_element(ptr.right, temp.data);
            }

            return ptr;
        }
    }

}

File: Swift/Structures/BST.swift
final class Node{
    var left:Node? = nil;
    var right:Node? = nil;
    lazy var data:Int;
    
    init(data:Int) {
        self.data = data;
    }
}


class BST{
    private var m_root:Node? =  nil;
    private var m_size:Int = 0;

    func insert(data:Int){
        let node:Node = Node(data);

        if self.is_empty(){
            self.m_root = node;
        }
        else{
            var ptr:Node? = self.m_root;

            while(ptr?.left != nil || ptr?.right != nil){
                if (ptr!.data > data){ ptr = ptr?.left;}
                else if (ptr!.data < data) { ptr = ptr?.right;}
                else{ return;}
            }

            if (ptr!.data > data){ ptr?.left = node;}
            else { ptr?.right = node; }
        }
    }

    func remove(data:Int, node:Node? = self.m_root) -> Node{
        if(node == nil){ return  node;}
        else if (data < node!.data) {
            node.left = self.remove(data, node?.left);
        }
        else if (data > node!.data) {
            node.right = self.remove(data, node?.right);
        }
        else{
            // case 1:
            if (node?.left == nil and node?.right == nil){node = nil}
            // case 2:
            else if (node?.left == nil) {
                var ptr:Node? = node;
                node = node?.right;
                ptr = nil; // will it work as expected?
            }
            else if (node?.right == nil) {
                var ptr:Node? = node;
                node = node?.left;
                ptr = nil; // will it work as expected?
            }
            // case 3:
            else{
                var ptr:Node? = self.find_min(node);
                node!.data = ptr!.data;
                ptr?.right = self.remove(ptr!.data, node?.right);
            }
        }

    }

    func traverse_preorder(root:Node = self.m_root){
        if(root != nil){
            print(" \(root!.data)");            
            traverse_preorder(root: self.m_root?.left);
            traverse_preorder(root: self.m_root?.right);
        }
    }

    func traverse_postorder(root:Node = self.m_root){
        if(root != nil){
            traverse_postorder(root: self.m_root?.left);
            traverse_postorder(root: self.m_root?.right);
            print(" \(root!.data)");            
        }
    }

    func traverse_inorder(root:Node = self.m_root){
        if(root != nil){
            traverse_inorder(root: self.m_root?.left);
            print(" \(root!.data)");            
            traverse_inorder(root: self.m_root?.right);
        }
    }

    func is_empty()->Bool{
        return self.m_root == nil;
    }

    func find_min(node:Node)->Node{
        while(node.left != nil){
            node = node.left;
        }
        return node;
    }

}